OpenRA.Test/OpenRA.Game/SyncImplTest.cs

```diff
@@ -0,0 +1,71 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2015 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation. For more information,
+ * see COPYING.
+ */
+#endregion
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Reflection;
+using NUnit.Framework;
+
+namespace OpenRA.Test
+{
+	[TestFixture]
+	class SyncImplTest
+	{
+		static readonly IEnumerable<Type> AllTypes =
+			Directory.EnumerateFiles("mods", "OpenRA.Mods.*.dll", SearchOption.AllDirectories)
+			.Append("OpenRA.Game.exe")
+			.Select(path => Assembly.LoadFile(Path.GetFullPath(path)))
+			.SelectMany(assembly => assembly.GetTypes());
+
+		static readonly Type SyncInterface = typeof(ISync);
+
+		static bool TypeImplementsSync(Type type)
+		{
+			// WTF: return type.GetInterfaces().Contains(SyncInterface) requires System.Drawing to be referenced!
+			foreach (var interfaceType in type.GetInterfaces())
+				if (interfaceType == SyncInterface)
+					return true;
+			return false;
+		}
+
+		static bool AnyTypeMemberIsSynced(Type type)
+		{
+			const BindingFlags Flags = BindingFlags.Public | BindingFlags.NonPublic | BindingFlags.Instance | BindingFlags.DeclaredOnly;
+			while (type != null)
+			{
+				if (((MemberInfo[])type.GetFields(Flags)).Concat(type.GetProperties(Flags)).Any(x => x.HasAttribute<SyncAttribute>()))
+					return true;
+				type = type.BaseType;
+			}
+
+			return false;
+		}
+
+		[TestCase(TestName = "Types with syncable members implement sync interface")]
+		public void TypesWithSyncableMembersImplementSyncInterface()
+		{
+			foreach (var type in AllTypes)
+				if (!TypeImplementsSync(type) && AnyTypeMemberIsSynced(type))
+					Assert.Fail("{0} has members with the Sync attribute but does not implement ISync".F(type.FullName));
+			Assert.Pass();
```

Comment:
http://www.nunit.org/index.php?p=utilityAsserts&r=2.5

> The Assert.Pass method allows you to immediately end the test, recording it as successful. Since it causes an exception to be thrown, it is more efficient to simply allow the test to return.